homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,683 @@
+            else:
+                self._state = "off"
+
+        self._name = (
+            self._backyard["BackyardName"]
+            + " "
+            + self.bow.get("Name")
+            + " "
+            + self.entitydata.get("Name")
+            + " Setting"
+        )
+
+        super().add_alarms(sensordata.get("Alarms"))
+
+        return self._state
+
+
+class OmniLogicPHSensor(OmnilogicSensor):
+    """Define an OmniLogic pH Sensor."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        backyard: dict,
+        bow: dict,
+        device_class: str,
+        icon: str,
+        unit: str,
+        sensordata: dict,
+    ):
+        """Initialize the sensor."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            device_class=device_class,
+            icon=icon,
+            unit=unit,
+            sensordata=sensordata,
+        )
+
+    @property
+    def state(self):
+        """Return the state for the pH sensor."""
+        _LOGGER.debug("Updating state of sensor: %s", self._name)
+
+        sensordata = {}
+
+        for backyard in self.coordinator.data:
+            for bow in backyard.get("BOWS"):
+                if bow["CSAD"]["systemId"] == self.entitydata["systemId"]:
+                    sensordata = bow.get("CSAD")
+                    break
+
+        phstate = None
+        if sensordata.get("ph") != 0:
+            phstate = sensordata.get("ph")
+
+        self._state = phstate
+        self._unit = "pH"
+        self._name = self._backyard["BackyardName"] + " " + self.bow.get("Name") + " pH"
+
+        super().add_alarms(sensordata.get("Alarms"))
+
+        return self._state
+
+
+class OmniLogicORPSensor(OmnilogicSensor):
+    """Define an OmniLogic ORP Sensor."""
+
+    def __init__(
+        self,
+        coordinator: OmniLogicUpdateCoordinator,
+        kind: str,
+        name: str,
+        backyard: dict,
+        bow: dict,
+        device_class: str,
+        icon: str,
+        unit: str,
+        sensordata: dict,
+    ):
+        """Initialize the sensor."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            device_class=device_class,
+            icon=icon,
+            unit=unit,
+            sensordata=sensordata,
+        )
+
+    @property
+    def state(self):
+        """Return the state for the ORP sensor."""
+        _LOGGER.debug("Updating state of sensor: %s", self._name)
+
+        sensordata = {}
+
+        for backyard in self.coordinator.data:
+            for bow in backyard.get("BOWS"):
+                if bow["CSAD"]["systemId"] == self.entitydata["systemId"]:
+                    sensordata = bow.get("CSAD")
+                    break
```

Comment:
Thanks for splitting this. However, I'm seeing a lot of repeated code (such as this block), which makes me think that either there's a better way to split them, or you can extract some work into methods on the base class.

Reply:
Unfortunately due to the structure of the data returned by OmniLogic this section varies slightly by sensor type and the identifiers and location  of the correct sensor need to be searched this way. The same sensor type may appear in multiple instances and different levels of the hierarchy and it is specific to the sensor.

Reply:
That's a bummer. I am seeing some duplication though (like the code I highlighted here), that can be extracted into methods.

Reply:
I'll have a look. Each sensor comparison for the final if statement identifying the correct telemetry data from the payload returned from Omnilogic is slightly different. Not sure how I can strip this any further with my Python knowledge.

Reply:
This code, for example, is identical to the beginning of the `OmniLogicPHSensor` implementation. I believe I saw a few other duplications.

Reply:
OK - found a couple I could compress into find_chlorinator and find_csad. The rest I don't think I can gain any efficiency on since they differ significantly. Pulling them out would just create a more complex logic loop somewhere else.